### ISSUE TITLE
Performance degradation after distinct pushdown

### ISSUE THREAD

**Issue by aalexandrov** [MaterializeInc/materialize#10119]
### What version of Materialize are you using?

v0.15.0 (f79f63205)

### How did you install Materialize?

Built from source

### Environment

```sql
-- database
CREATE DATABASE reduction_pushdown;
-- schema
CREATE TABLE R(x INT NOT NULL, y INT NOT NULL, z INT NOT NULL);
-- data
INSERT INTO R VALUES
    (1, 1, 1),
    (1, 2, 1),
    (1, 3, 1),
    (2, 4, 1),
    (2, 5, 1),
    (3, 6, 1),
    (3, 7, 0);
```

### Queries

```sql
SELECT DISTINCT
    R.x,
    R.y,
    R.z,
    R.x / R.z
FROM
    R
INNER JOIN 
    (VALUES (1),(2)) AS S(x) ON R.x = S.x
```

### Plans

<table>
  <thead>
    <tr>
      <td><b>Optimized Plan (0.12.0)</b></td>
      <td><b>Optimized Plan (0.15.0)</b></td>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><img src="https://user-images.githubusercontent.com/1071946/150112390-5380a659-844d-4716-b571-11684b882cf5.png" /></td>
      <td><img src="https://user-images.githubusercontent.com/1071946/150112382-01f1039b-8506-4fb4-a6bd-16c5012fdabf.png" /></td>
    </tr>
  </tbody>
</table>

### Problem Discussion

The following is an example of two regressions observed for the given query and environment between versions `0.12.0` and `0.15.0`.

1. The query fails with a `division by zero` error.
2. The query consumes memory proportional to the number of distinct `R.x` values and not the number of `S(x)` values used to restrict `R`.

Both issues are caused by the rewrite introduced in #8399 which pushes `Distinct` through the `Join` that implements the list-based filter. 

Regression (1) is not really a bug, because the SQL standard does not guarantee the order of execution of filters. Rather, this is an example of an issue with the user query which does not really protect itself against "division by zero", so I suggest to offload these type of fixes to the use.

Regression (2) is more concerning, because the number of distinct values in `R.x` might be much bigger than `S.x`. I therefore suggest to prohibit reduction pushdown in cases where the other side of the join is a literal collection with size up to a certain point.

**Comment by teskje** [comment]
We still want to push down the reduction to constant inputs, to be able to fuse them away. For this, we can perform what's referred to as a "partial pushdown", i.e., pushing down the reduction to only some of the inputs.

For example:
```
Get r         Constant (1) (1) (2)
  | %0               | %1
  --------------------
           |
   Join %0 %1 (= #0 #1)
   Distinct group=(#0)
```

... can be transformed to:
```
Get r         Constant (1) (1) (2)
  |           Distinct group=(#0)
  | %0               | %1
  --------------------
           |
   Join %0 %1 (= #0 #1)
   Distinct group=(#0)
```

... and the pushed-down `Distinct` can be fused with the `Constant` in a subsequent step.

The general strategy here is: If there are constant relations in the join inputs, perform a modified version of our current reduction pushdown that skips pushing down the reduction to non-constant inputs and leaves the reduction around the join in place.

Note that this strategy might not be sound for all kinds of aggregation functions, but we currently only support pushdown of `Distinct`s, for which it should be sound.

**Comment by teskje** [comment]
The assumption that the approach outlined above is correct for all distinct aggregations is wrong: It doesn't work for `count` at least:

```
Get r         Constant (1, 2) (1, 3)
  | %0               | %1
  --------------------
           |
   Join %0 %1 (= #0 #1)
   Reduce group=(#1)
    - agg count(distinct #2)
```

... would become:

```
Get r         Constant (1, 2) (1, 3)
  |               Reduce group=(#0)
  |                - agg count(distinct #1)
  | %0               | %1
  --------------------
           |
   Join %0 %1 (= #0 #1)
   Reduce group=(#1)
    - agg count(distinct #2)
```

For example, if `r` contains a single record `(1)`, then the result would be `(1, 2)` in the original case, but `(1, 1)` in the transformed case.

**Comment by teskje** [comment]
@aalexandrov pointed my to https://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.63.6449&rep=rep1&type=pdf, which describes the possibility to decompose aggregations and then pushing down the "inner" aggregation while replacing the original aggregation with the "outer" one. Based on this, my example above should probably use `sum` instead of `count` in the reduce around the join.

**Comment by aalexandrov** [comment]
Let's just handle the `distinct` only case in this issue and open a follow-up to rethink the reduction push-down at some point in the future.

**Comment by wangandi** [comment]
In the examples below, `r` has two columns instead of one.

The way I see it, there are two cases for partial pushdown in the case of single-component distinct aggregations:

(1)The aggregation involves a column in the non-constant component:

```
Get r         Constant (1, 2) (1, 3)
  | %0               | %1
  --------------------
           |
   Join %0 %1 (= #0 #2)
   Reduce group=(#2)
    - agg count(distinct #1)
```
In this case, we should be able to just push a distinct down to the constant without affecting the result of the distinct aggregation.
```
Get r         Constant (1, 2) (1, 3)
  |            Distinct group=(#0)
  | %0               | %1
  --------------------
           |
   Join %0 %1 (= #0 #1)
   Reduce group=(#2)
    - agg count(distinct #1)
```

(2) The aggregation involves a column in the constant component:
```
Get r         Constant (1, 2) (1, 3)
  | %0               | %1
  --------------------
           |
   Join %0 %1 (= #0 #2)
   Reduce group=(#2)
    - agg count(distinct #3)
```
In this case, after pushing down the aggregation, we just need to get rid of additional copies of rows created as a result of the join.
```
Get r         Constant (1, 2) (1, 3)
  |               Reduce group=(#0)
  |               - agg count(distinct #1)
  | %0               | %1
  --------------------
           |
   Join %0 %1 (= #0 #2)
   Distinct (#2, #3)
```

**Comment by uce** [comment]
Thank you all for your analysis. The assumption for this ticket was that it is a low-risk good first issue. I don't think that's actually the case, given the above discussion.

Given our ongoing push for the next milestone and the risk of introducing a bug here, I would like us to unassign this ticket and leave it for another time. It's not a great experience for Jan, but I think he has gotten a good impression of the optimizer (which was the original goal why he picked it up).

**Comment by aalexandrov** [comment]
I think the [case from this comment](https://github.com/MaterializeInc/materialize/issues/10119#issuecomment-1098143439) should be good to be merged. 

For [the extra cases proposed yesterday](https://github.com/MaterializeInc/materialize/issues/10119#issuecomment-1098143439) I think we should punt and use the recently proposed template. It is hard for me to follow under which assumptions we are operating and what mathematical properties are we using to justify the correctness of these rewrites. I feel like we have a document where these are clearly defined, so people can use it as a primer when thinking about concrete rewrites.

**Comment by teskje** [comment]
We decided in the optimizer daily that we should handle the partial reduction pushdown optimization the proper way, by following the template process @aalexandrov mentioned above. I have closed my draft PR (#11798) and opened another one to salvage the unit tests (#11845), as they might come in handy later.

@wangandi wanted to create a new issue for the partial reduction pushdown feature. Once that exists, we can close this one as a duplicate.
